include/devicetree.h

```diff
@@ -1,15 +1,1283 @@
+ *
+ * For properties whose bindings have the following types, this macro
+ * expands to:
+ *
+ * - string: a string literal
+ * - boolean: 0 if the property is false, or 1 if it is true
+ * - int: the property's value as an integer literal
+ * - array, uint8-array, string-array: an initializer expression in braces,
+ *   whose elements are integer or string literals (like {0, 1, 2},
+ *   {"hello", "world"}, etc.)
+ * - phandle: a node identifier
+ *
+ * For other properties, behavior is undefined.
+ *
+ * For examples, see @ref DT_PATH(), @ref DT_ALIAS(), @ref DT_NODELABEL(),
+ * and @ref DT_INST().
+ *
+ * @param node_id node identifier
+ * @param prop lowercase-and-underscores property name
+ * @return a representation of the property's value
+ */
+#define DT_PROP(node_id, prop) DT_CAT(node_id, _P_##prop)
+
+/**
+ * @brief Get a property's logical length
+ *
+ * Here, "length" is a number of elements, which may not
+ * be a size in bytes.
+ *
+ * For properties whose binding has type array, string-array, or
+ * uint8-array, this expands to the number of elements in the array.
+ *
+ * For properties of type phandles or phandle-array, it expands to the
+ * number of phandles or phandle+specifiers respectively.
+ *
+ * These properties are handled as special cases:
+ *
+ * - reg property: use DT_NUM_REGS(node_id) instead
+ * - interrupts property: use DT_NUM_IRQS(node_id) instead
+ *
+ * It is an error to use this macro with the above properties.
+ *
+ * For other properties, behavior is undefined.
+ *
+ * @param node_id node identifier
+ * @param prop a lowercase-and-underscores property with a logical length
+ * @return the property's length
+ */
+#define DT_PROP_LEN(node_id, prop) DT_PROP(node_id, prop##_LEN)
+
+/**
+ * @brief Is index "idx" valid for an array type property?
+ *
+ * If this returns 1, then DT_PROP_IDX(node_id, prop, idx) or
+ * DT_PHA_IDX(node_id, pha, idx, cell) are valid at index "idx".
+ * If it returns 0, it is an error to use those macros with that index.
+ *
+ * These properties are handled as special cases:
+ *
+ * - reg property: use DT_REG_HAS_IDX(node_id, idx) instead
+ * - interrupts property: use DT_IRQ_HAS_IDX(node_id, idx) instead
+ *
+ * @param node_id node identifier
+ * @param prop a lowercase-and-underscores property with a logical length
+ * @param idx index to check
+ * @return 1 if "idx" is a valid index into the given property,
+ *         0 otherwise.
+ */
+#define DT_PROP_HAS_IDX(node_id, prop, idx) \
+	((idx) < DT_PROP_LEN(node_id, prop))
+
+/**
+ * @brief Get the value at index "idx" in an array type property
+ *
+ * It might help to read the argument order as being similar to
+ * "node->property[index]".
+ *
+ * When the property's binding has type array, string-array,
+ * uint8-array, or phandles, this expands to the idx-th array element
+ * as an integer, string literal, or node identifier respectively.
+ *
+ * It's not safe to use this macro with the special "reg" or
```

Comment:
This is the third way a special case for `reg` and `interrupts` is identified and described.  That's not wrong, but some people might see the differences and wonder if they're significant....

Reply:
I think it's wrong. Will fix; thanks.